src/puppetlabs/puppetdb/cli/services.clj

```diff
@@ -193,26 +191,33 @@
   [{:keys [broker mq-factory mq-connection]}]
   (when broker
     (log/info "Shutting down the messsage queues.")
-    (.close mq-connection)
-    (.close mq-factory)
+    (doseq [conn [mq-connection mq-factory]]
+      (.close conn))
     (mq/stop-broker! broker)))
 
+(defn shutdown-updater
+  [{:keys [updater]}]
+  (when updater
+    (log/info "Shutting down updater thread.")
+    (future-cancel updater)))
+
+(defn shutdown-db-connections
+  [{{read-db :scf-read-db
+     write-db :scf-write-db} :shared-globals}]
+  (doseq [db [write-db read-db]]
+    (when-let [ds (:datasource db)] (.close ds)))) 
+
 (defn stop-puppetdb
   "Shuts down PuppetDB, releasing resources when possible.  If this is
   not a normal shutdown, emergency? must be set, which currently just
   produces a fatal level level log message, instead of info."
-  [context emergency?]
+  [context & [emergency?]]
   (if emergency?
     (log/error "A fatal error occurred; shutting down all subsystems.")
     (log/info "Shutdown request received; puppetdb exiting."))
-  (when-let [updater (context :updater)]
-    (log/info "Shutting down updater thread.")
-    (future-cancel updater))
   (shutdown-mq context)
-  (when-let [ds (get-in context [:shared-globals :scf-write-db :datasource])]
-    (.close ds))
-  (when-let [ds (get-in context [:shared-globals :scf-read-db :datasource])]
-    (.close ds))
+  (shutdown-updater context)
```

Comment:
you've changed the shutdown ordering here -- I don't think it matters but was there a reason for that?